snips_nlu/preprocessing.py

```diff
@@ -1,14 +1,83 @@
+# coding=utf-8
 from __future__ import unicode_literals
 
+from builtins import object
+
+from snips_nlu_utils import (tokenize as _tokenize,
+                             tokenize_light as _tokenize_light,
+                             normalize)
+
 from snips_nlu.resources import get_stems
-from snips_nlu.tokenization import tokenize_light
 
 
 def stem(string, language):
     tokens = tokenize_light(string, language)
     stemmed_tokens = [_stem(token, language) for token in tokens]
-    return ' '.join(stemmed_tokens)
+    return " ".join(stemmed_tokens)
 
 
 def _stem(string, language):
     return get_stems(language).get(string, string)
+
+
+class Token(object):
+    """Token object which is output by the tokenization
+
+    Attributes:
+        value (str): Tokenized string
+        normalized (str): Normalized value of the tokenized string
+        start (int): Start position of the token within the sentence
+        end (int): End position of the token within the sentence
+    """
+
+    def __init__(self, value, start, end):
+        self.value = value
+        self.start = start
+        self.end = end
+        self._normalized_value = None
+        self._stem = None
+
+    def get_normalized_value(self):
+        if self._normalized_value is not None:
+            return self._normalized_value
+        self._normalized_value = normalize(self.value)
+        return self._normalized_value
+
+    def get_stem(self, language):
```

Comment:
It feels super weird to have to pass the language to access what really is an attribute of the token.
We should make it a property initialized to `None`.
Then when someone expects tokens to be stemmed it uses the function `tokens = [stem(t, language) for t in tokens]` and it's actually the stem function that set the `stem` attribute to something which is not `None`.
I find it more elegant this way. WDYT?